File: src/scanners/osv.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import type { IScanner, ImageSource, ScannerResult } from '../types/index.js';
import { execWithTimeout, getToolVersion, isToolAvailable, formatSourceRef, parseJsonFile, writeFallbackResult, createTimedResult } from './utils.js';

const TIMEOUT_MS = 300_000;

export class OsvScanner implements IScanner {
  readonly name = 'osv';

  async scan(source: ImageSource, outputPath: string, sbomPath?: string): Promise<ScannerResult> {
    const start = Date.now();
    const reportDir = path.dirname(outputPath);
    const ownSbom = sbomPath ?? path.join(reportDir, 'osv-sbom.cdx.json');
    let cleanupSbom = !sbomPath;

    try {
      // Generate SBOM if not provided (reuse Syft output when available)
      if (!sbomPath) {
        const syftSbom = path.join(reportDir, 'sbom.cdx.json');
        try {
          await fs.access(syftSbom);
          // Reuse existing Syft SBOM
          await fs.copyFile(syftSbom, ownSbom);
          cleanupSbom = true;
        } catch {
          // Generate independent SBOM
          const ref = formatSourceRef(source);
          await execWithTimeout(
            `syft ${ref} -o cyclonedx-json@1.5 > "${ownSbom}"`,
            TIMEOUT_MS,
          );
        }
      }

      // OSV scanner exits code 1 when vulns are found — that is success
      try {
        await execWithTimeout(
          `osv-scanner -L "${ownSbom}" --verbosity error --format json > "${outputPath}"`,
          TIMEOUT_MS,
          { maxBuffer: 100 * 1024 * 1024 },
        );
      } catch (osvError) {
        // Check if output file was written (exit code 1 = vulns found)
        try {
          await fs.access(outputPath);
        } catch {
          throw osvError;
        }
      }

      const data = await parseJsonFile(outputPath);
      return createTimedResult(this.name, true, start, { data });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.warn('OSV scan failed:', message);
      await writeFallbackResult(outputPath, message, { vulnerabilities: [] });
      return createTimedResult(this.name, false, start, { error: message });
    } finally {
      if (cleanupSbom) {
        try {
          await fs.unlink(ownSbom);
        } catch {
          // Ignore cleanup errors
        }
      }
    }
  }

  async getVersion(): Promise<string> {
    return getToolVersion('osv-scanner --version');
  }

  async isAvailable(): Promise<boolean> {
    return isToolAvailable('osv-scanner');
  }

  supportsSource(_source: ImageSource): boolean {
    return true;
  }
}

File: internal/scanner/osv.go
package scanner

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

const osvTimeoutMs = 300000

type OsvScanner struct{}

func (o *OsvScanner) Name() string { return "osv" }

func (o *OsvScanner) Scan(ctx context.Context, source types.ImageSource, outputPath string) (*types.ScannerResult, error) {
	start := time.Now()

	reportDir := filepath.Dir(outputPath)
	ownSbom := filepath.Join(reportDir, "osv-sbom.cdx.json")
	cleanupSbom := true

	// Try to reuse Syft's SBOM if available
	syftSbom := filepath.Join(reportDir, "sbom.cdx.json")
	if _, err := os.Stat(syftSbom); err == nil {
		// Copy the existing SBOM
		data, readErr := os.ReadFile(syftSbom)
		if readErr == nil {
			_ = os.WriteFile(ownSbom, data, 0644)
		}
	} else {
		// Generate independent SBOM
		ref := FormatSourceRef(source.Type, source.Ref, source.Path)
		cmd := fmt.Sprintf(`syft %s -o cyclonedx-json@1.5 > "%s"`, ref, ownSbom)
		var sbomEnv []string
		if source.Insecure && source.Type == "registry" {
			sbomEnv = BuildEnv(map[string]string{
				"SYFT_REGISTRY_INSECURE_SKIP_TLS_VERIFY": "true",
				"SYFT_REGISTRY_INSECURE_USE_HTTP":        "true",
			})
		}
		stdout, stderr, err := ExecWithTimeout(ctx, cmd, osvTimeoutMs, sbomEnv)
		if err != nil {
			msg := err.Error()
			if ctx.Err() != nil {
				msg = "scan cancelled"
			} else {
				if s := strings.TrimSpace(stderr); s != "" {
					msg += "\n--- syft (osv sbom) stderr ---\n" + s
				}
				if s := strings.TrimSpace(stdout); s != "" {
					msg += "\n--- syft (osv sbom) stdout ---\n" + s
				}
			}
			fmt.Fprintf(os.Stderr, "OSV scan failed: %s\n", msg)
			_ = WriteFallbackResult(outputPath, msg, map[string]interface{}{"vulnerabilities": []interface{}{}})
			durationMs := time.Since(start).Milliseconds()
			return &types.ScannerResult{Scanner: "osv", Success: false, Error: msg, DurationMs: durationMs}, nil
		}
	}

	// Run osv-scanner — exit code 1 means vulns found (success)
	cmd := fmt.Sprintf(`osv-scanner -L "%s" --verbosity error --format json > "%s"`, ownSbom, outputPath)
	stdout, stderr, err := ExecWithTimeout(ctx, cmd, osvTimeoutMs, nil)

	if err != nil {
		// Check if output file was written (exit code 1 = vulns found)
		if _, statErr := os.Stat(outputPath); statErr != nil {
			msg := err.Error()
			if s := strings.TrimSpace(stderr); s != "" {
				msg += "\n--- osv-scanner stderr ---\n" + s
			}
			if s := strings.TrimSpace(stdout); s != "" {
				msg += "\n--- osv-scanner stdout ---\n" + s
			}
			fmt.Fprintf(os.Stderr, "OSV scan failed: %s\n", msg)
			_ = WriteFallbackResult(outputPath, msg, map[string]interface{}{"vulnerabilities": []interface{}{}})
			if cleanupSbom {
				_ = os.Remove(ownSbom)
			}
			durationMs := time.Since(start).Milliseconds()
			return &types.ScannerResult{Scanner: "osv", Success: false, Error: msg, DurationMs: durationMs}, nil
		}
		// Output exists — vulns found is success
	}

	if cleanupSbom {
		_ = os.Remove(ownSbom)
	}

	var data interface{}
	if parseErr := ParseJSONFile(outputPath, &data); parseErr != nil {
		msg := parseErr.Error()
		durationMs := time.Since(start).Milliseconds()
		return &types.ScannerResult{Scanner: "osv", Success: false, Error: msg, DurationMs: durationMs}, nil
	}

	durationMs := time.Since(start).Milliseconds()
	return &types.ScannerResult{Scanner: "osv", Success: true, Data: data, DurationMs: durationMs}, nil
}

func (o *OsvScanner) GetVersion() string {
	return GetToolVersion("osv-scanner --version")
}

func (o *OsvScanner) IsAvailable() bool {
	return IsToolAvailable("osv-scanner")
}

func (o *OsvScanner) SupportsSource(_ types.ImageSource) bool {
	return true
}
